server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java

```diff
@@ -280,20 +288,38 @@ public boolean isLeader()
       final List<Rule> rules = metadataRuleManager.getRulesWithDefault(segment.getDataSource());
 
       for (final Rule rule : rules) {
-        if (!(rule instanceof LoadRule && rule.appliesTo(segment, now))) {
+        if (!rule.appliesTo(segment, now)) {
           continue;
         }
 
-        ((LoadRule) rule)
-            .getTieredReplicants()
-            .forEach((final String tier, final Integer ruleReplicants) -> {
-              int currentReplicants = segmentReplicantLookup.getLoadedReplicants(segment.getId(), tier);
-              Object2LongMap<String> underReplicationPerDataSource = underReplicationCountsPerDataSourcePerTier
-                  .computeIfAbsent(tier, ignored -> new Object2LongOpenHashMap<>());
+        if (rule instanceof LoadRule) {
+          ((LoadRule) rule)
+              .getTieredReplicants()
+              .forEach((final String tier, final Integer ruleReplicants) -> {
+                int currentReplicants = segmentReplicantLookup.getLoadedReplicants(segment.getId(), tier);
+                Object2LongMap<String> underReplicationPerDataSource = underReplicationCountsPerDataSourcePerTier
+                    .computeIfAbsent(tier, ignored -> new Object2LongOpenHashMap<>());
+                ((Object2LongOpenHashMap<String>) underReplicationPerDataSource)
+                    .addTo(segment.getDataSource(), Math.max(ruleReplicants - currentReplicants, 0));
+              });
+        }
+
+        if (rule instanceof BroadcastDistributionRule) {
```

Comment:
If `Rule` subclasses are added in the future and should be considered in this method, is there a test that will fail?

Reply:
Not right now. A Rule subclass may not always be needed to be considered in this method. Also not sure how the test will be able to automatically create new Rule subclass

Reply:
Maybe adding some comments to somewhere like `Rule` will be sufficient for now. Not sure how likely we'll add future `Rule`s, but if we do I think there's a good chance we'll forget to update this method if it's needed.

Reply:
https://github.com/apache/druid/pull/10054